iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -241,19 +264,29 @@ class ProcessTransferWrite final
   }
 };
 
-static Optional<MemRefType> getVectorizedMemRefType(
-    ConversionPatternRewriter &rewriter, MemRefType type) {
+template <typename OpTy>
+Optional<MemRefType> MemRefConversionPattern<OpTy>::getVectorizedMemRefType(
+    ConversionPatternRewriter &rewriter, Value memRefValue) const {
+  unsigned vecSizeInBits =
+      memrefUsageAnalysis.getMemRefVectorSizeInBits(memRefValue);
+  MemRefType type = memRefValue.getType().cast<MemRefType>();
   unsigned elemSize = type.getElementTypeBitWidth();
-  unsigned vecSize = kVectorizationSizeInBits / elemSize;
-  // Pick a new type of element size 32bits.
-  Type newElemType = type.getElementType().isa<IntegerType>()
-                         ? rewriter.getI32Type().cast<Type>()
-                         : rewriter.getF32Type().cast<Type>();
-  Type vecType = VectorType::get(kVecSize, newElemType);
+  unsigned numElements = vecSizeInBits / elemSize;
+  Type elemType = type.getElementType();
+  // If the vector we need to generate is bigger than the the max vector size
+  // allowed for loads use a larger element type.
+  if (numElements > kMaxVectorNumElements) {
```

Comment:
I am actually not sure what is happening here. Based on the number of elements, the type is changing. That is counter-intuitive to me.

Reply:
Yes, the idea is that before we were trying to convert everything to memref<..vec<4xi32>>, this change is keeping the original type if the vectorization size we want is smaller. For instance if we are loading vec<4xf16>, the memref would get converted to memref<..vec<4xf16>> but if we are trying load vec<8xf16> this will create a memref<...vec<4xf32>> and cast the result of the load to vec<4xf16>. This can useful because sometimes we a type smaller than 4bytes and we still want to load it as a larger type and cast it to get better memory access patterns. 
I can add more comments if you think that is what is missing or we can chat more about the logic.

Reply:
Why cant we convert to `memref<...<vec<8xf16>>`?

It might be easier to this in the analysis itself. Figure out the memref type to use for the allocation, and adjust the loads appropriately?

Let me know if you want to chat offline?

Reply:
> Why cant we convert to `memref<...<vec<8xf16>>`?
> 
> It might be easier to this in the analysis itself. Figure out the memref type to use for the allocation, and adjust the loads appropriately?
> 
> Let me know if you want to chat offline?

I added a comment to explain the logic. Hopefully that makes it more clear.